lib/tarfiles.py

```diff
@@ -128,50 +135,101 @@ def do_tarballs(args):
-    else:
-        res = requests.post(url, cert=(proxy, proxy), data=tf, verify=False)
-    if res.text[:8] == "PRESENT:":
-        return res.text[8:]
-    else:
-        return None
-
-
-def pubapi_exists(cid, proxy = None, tokenstr = None):
-    """make pubapi update call to check if we already have this tarfile,
-    return path.
+class TarfilePublisherHandler(object):
+    """Handler to publish tarballs via HTTP to RCDS (or future dropbox server)
+
+    Args:
+        object (_type_): _description_
+        cid (str): unique group/hash combination that RCDS uses to locate tarballs
+        proxy (str): Location of X509 Proxy file to authenticate to RCDS
+        token (str): Location of JWT/Sci-token to authenticate to RCDS
     """
+
     dropbox_server = "rcds.fnal.gov"
-    url = "https://%s/pubapi/exists?cid=%s" % (dropbox_server, cid)
-    if tokenstr:
-        headers = { 'Authorization': 'Bearer %s' % tokenstr}
-        res = requests.get(url, headers=headers , verify=False)
-    else:
-        res = requests.get(url, cert=(proxy, proxy), verify=False)
-    if res.text[:8] == "PRESENT:":
-        return res.text[8:]
-    else:
-        return None
+    pubapi_base_url = f"https://{dropbox_server}/pubapi"
+    check_tarball_present_re = re.compile(
+        "^PRESENT\:(.+)$"
+    )  # RCDS returns this if a tarball represented by cid is present
+
+    def __init__(self, cid, proxy=None, token=None):
+        self.cid_url = _quote(cid, safe="")  # Encode CID for passing to URL
+        self.proxy = proxy
+        self.token = token
+        self.pubapi_base_url_formatter = (
+            f"{self.pubapi_base_url}/{{endpoint}}?cid={self.cid_url}"
+        )
+        if token is not None:
+            self.request_headers = self.__make_request_token_headers()
+            print(f"Using bearer token located at {self.token} to authenticate to RCDS")
+        else:
+            print(f"Using X509 proxy located at {self.proxy} to authenticate to RCDS")
+
+
+    # Some sort of wrapper to wrap the above three
+    def pubapi_operation(func):
+        """Wrap various PubAPI operations, return path if we get it from response"""
+
+        def wrapper(self, *args, **kwargs):
+            response = func(self, *args, **kwargs)
+            _match = self.check_tarball_present_re.match(response.text)
+            if _match is not None:
+                return _match.group(1)
+            return None
+
+        return wrapper
+
+    @pubapi_operation
+    def update_cid(self):
+        """Make PubAPI update call to check if we already have this tarfile
+
+        Returns:
+            requests.Response: Response from PubAPI call indicating if tarball
+            represented by self.cid is present
+        """
+        url = self.pubapi_base_url_formatter.format(endpoint="update")
+        if self.token:
+            return requests.get(url, headers=self.request_headers, verify=False)
```

Comment:
What would it take to be able to verify the server cert?

Reply:
The problem with that is that the RCDS hostcerts don't seem to have `rcds.fnal.gov` as SANs, so the verify bit fails when we try to connect to `rcds.fnal.gov`.  I'll talk to Dave and Hyunwoo about that to see if they're OK with switching those certs out.

Reply:
Talked to Dave and Kevin, and we all agreed that mimicking the old jobsub behavior is best - that is choosing between rcds01 and rcds02 randomly, then if it fails, using the other one.  We don't expect RCDS to expand, and if it does, we have the env var already we can use to override.

Reply:
Done.